File: Cpp/821 Shortest distance to a character.cpp
#include <gtest\gtest.h>
#include <numeric>
#include <queue>

using namespace std;

vector<int> shortestToChar(string S, char C) 
{	
	queue<int> position;
	for (int i = 0; i < S.length(); ++i)
	{
		if (S[i] == C)
		{
			position.push(i);
		}
	}
		
	int next = position.front();
	int begin = -next;
	position.pop();

	vector<int> result;
	for (int i = 0; i < S.length(); ++i)
	{
		if (i > next)
		{		
			begin = next;
			if (!position.empty())
			{				
				next = position.front();
				position.pop();
			}
			else
			{
				next = numeric_limits<int>::max();
			}			
		}

		result.push_back(std::min(i - begin, next - i));
	}
	
	return result;
}

TEST(LeetCodeCN, tShortestToChar)
{
	auto result = shortestToChar("loveleetcode", 'e');
	result = shortestToChar("gtywtbcnowpwibfdgvph", 'i');
}

File: Python/Codes/Practice/LeetCodeCN/821 Shortest distance to a character.py

class Solution:
    def shortestToChar(self, S, C):
        """
        :type S: str
        :type C: str
        :rtype: List[int]
        """
        pos = []
        for i,v in enumerate(S):
            if v == C:
                pos.append(i)

        result = []
        begin = -20000
        next = pos.pop(0)
        for i in range(len(S)):
            if i > next:
                begin = next
                next = len(pos) and pos.pop(0) or 20000                

            result.append(min(i - begin, next-i))

        return result


if __name__ == '__main__':
    s = Solution()
    print(s.shortestToChar('gtywtbcnowpwibfdgvph','i'))
    print(s.shortestToChar('loveleetcode','e'))